### DISCUSSION TITLE
Implementing Resubmit for Auth behavior in WCF .NET Core client

### DISCUSSION THREAD

**Discussion by devprj22** [CoreWCF/CoreWCF#1392] [Q&A] [answered]
Hi,

We have an API that gets user status for other users.

The .NET Framework version of the API uses Negotiate and NTLM and the flow has resubmit logic, which resubmit the requests once server sends challenge string after the first request is made.

In the .NET Core implementation, we do not have this resubmit behavior and we are trying to implement it on our own using WCF to.

Here is the binding settings -

```
basicHttpBinding.Security.Mode = BasicHttpSecurityMode.Transport;
basicHttpBinding.Security.Transport.ClientCredentialType = HttpClientCredentialType.Windows;

ChannelFactory<WcfServiceProxy> channelFactory = new ChannelFactory<WcfServiceProxy>(basicHttpBinding, this.endpoint);
```

We followed the approach of implementing a `InterceptingHttpMessageHandler` using this blog post -

https://justsimplycode.com/2019/11/02/disable-header-100-continue-in-net-core-wcf-client/

Inside the SendAsync method, we check for the challenge string and then send the request with OAuth credentials -

The code looks like -

```
        protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
        {
            HttpResponseMessage response;
            response = await base.SendAsync(request, cancellationToken);

            try
            {
                if (response != null && response.StatusCode == HttpStatusCode.Unauthorized)
                {
                    IEnumerable<string> values;
                    if (response.Headers.TryGetValues(WellKnownHeaders.WWWAuthenticate, out values))
                    {
                        string challenge = values.First();

                        if (challenge != null && 
                            (   challenge.StartsWith("Bearer realm=", StringComparison.OrdinalIgnoreCase)
                            ||  challenge.StartsWith("Basic realm=", StringComparison.OrdinalIgnoreCase)))
                        {
                            Authorization authorization = new Authorization(challenge, request, clientCredentials);

                            if (authorization != null)
                            {
                                HttpRequestMessage cloneRequest = await CloneHttpRequestMessageAsync(request);

                                if (cloneRequest != null)
                                {
                                    // Add the authorization header.
                                    cloneRequest.Headers.Authorization = AuthenticationHeaderValue.Parse(authorization.Message);

                                    using (cloneRequest)
                                    {
                                        response.Dispose();

                                        HttpResponseMessage newResponse = await base.SendAsync(cloneRequest, cancellationToken);
                                        return newResponse;
                                    }
                                }
                            }
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                Log(ex);
                throw;
            }
            return response;
        }

```

The challenge string (present in WwwAuthenticate header) I get after the first call is -

"Basic Realm="

After I do the resubmit, I am still getting a 401.

If anyone knows what is missing in this, please share your insights.

**Comment by mconnew** [comment, marked as answer]
You probably need to have another HttpClientHandler instance that you create yourself that you embed inside your own handler. The one that WCF provides to you will have the Credentials property populated. I believe that will result in it replacing your auth header with its own. You need an inner HttpClientHandler that you use in the retry scenario with no credentials configured.  

Another point worth knowing, the WWW-Authenticate header can contain multiple auth schemes, so you're auth scheme might not be the first one which means the challenge you are interested in might not be at the start of the string. If you are in full control of the service, then you control that and you can choose for that to not be a problem. But if it's a 3rd party service, they could move it later in the header and your code might break.  

Another thing that might be causing you issues is where did you get the client credentials from (and what type are they). If you pulled it from the HttpClientHandler passed to your delegate, it's a CredentialCache and you need to supply the right url and auth scheme to get the credentials out of it. If the Authorization class takes an ICredentials instance, it might not be able to get to the NetworkCredential contained inside the CredentialCache. If that's the case, you can get the credentials out of the CredentialCache with the right magic incantation. Let me know if you need to do that and need help with the incantation.

**Reply by devprj22** [reply]
Thanks @mconnew, your suggestion related to properly parsing the WWW-Authenticate header instead of checking the specific scheme helped me fix it!

I had to do a first resubmit with empty auth header and re-submit again using the scheme received in the response of first challenge to properly get the auth token and get the requests working.

We were using our own credentials instead of pulling it in from the HttpClientHandler, so the remaining fix was not needed in this case.
